### PR TITLE
Update lucene version to 8.8.2

### PR DESCRIPTION
**Pull request by nknize**
This PR updates the codebase to the latest released version of Lucene. API changes are included and tested.

### CODE REVIEW

**Comment by nknize** [comment]
start gradle check

File: server/src/test/java/org/opensearch/index/codec/CodecTests.java
Comment (rmuir) [inline server/src/test/java/org/opensearch/index/codec/CodecTests.java:79]: Why do the stored fields now default to `BEST_COMPRESSION`? This is not an appropriate default: it should be `BEST_SPEED`. For DV `BEST_COMPRESSION` is appropriate here...

Comment (nknize) [reply server/src/test/java/org/opensearch/index/codec/CodecTests.java:79]: :+1:  Thanks @rmuir! This was an unintentional tab completion miss on my part.

Comment (rmuir) [reply server/src/test/java/org/opensearch/index/codec/CodecTests.java:79]: OK, the crazy compression options are a personal pet-peeve in lucene, (I added more comments to https://issues.apache.org/jira/browse/LUCENE-9843). Lucene nightly benchmark even accidentally uses wrong settings!

**Comment by nknize** [comment]
start gradle check

**Comment by peterzhuamazon** [comment]
start dco check
start wrapper validation
start gradle precommit

**Comment by nknize** [comment]
@rmuir do you want to give this another round of review?

File: server/src/main/java/org/opensearch/common/geo/GeoShapeUtils.java
Comment (mikemccand) [inline server/src/main/java/org/opensearch/common/geo/GeoShapeUtils.java:66]: Hmm why was this round trip encode/decode needed?

**Review by mikemccand — APPROVED** [review]
Looks great!

I love that we are aggressively upgrading to new Lucene releases.

I left one small question about the geo change, otherwise LGTM.

File: server/src/main/java/org/opensearch/common/geo/GeoShapeUtils.java
Comment (nknize) [reply server/src/main/java/org/opensearch/common/geo/GeoShapeUtils.java:66]: It's not... it's adding extra superfluous work! Reverting.. thx for the catch!

**Review by mikemccand — APPROVED** [review]

**Comment by nknize** [comment]
start gradle check

**Comment by nknize** [comment]
start gradle check

**Review by adnapibar — APPROVED** [review]
LGTM! 👍

File: server/src/main/java/org/opensearch/common/geo/GeoShapeUtils.java
Comment (nknize) [reply server/src/main/java/org/opensearch/common/geo/GeoShapeUtils.java]: Looks like a test failure as a result of reverting: 

```
REPRODUCE WITH: ./gradlew ':server:internalClusterTest' --tests "org.opensearch.search.geo.GeoShapeIntegrationIT.testIndexPolygonDateLine" -Dtests.seed=364ABAA5289793FA -Dtests.security.manager=true -Dtests.locale=de -Dtests.timezone=Africa/Ceuta -Druntime.java=15

org.opensearch.search.geo.GeoShapeIntegrationIT > testIndexPolygonDateLine FAILED
    java.lang.AssertionError: 
    Expected: <1L>
         but: was <0L>
        at __randomizedtesting.SeedInfo.seed([364ABAA5289793FA:DF31BD88058CBB72]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.junit.Assert.assertThat(Assert.java:923)
        at org.opensearch.search.geo.GeoShapeIntegrationIT.testIndexPolygonDateLine(GeoShapeIntegrationIT.java:300)
```

Something changed in upstream Lucene land that now requires quantization. I think we'll want to explore further. In the meantime I'll add a fix that doesn't require quantizing before indexing.

**Comment by nknize** [comment]
start gradle check

File: server/src/main/java/org/opensearch/common/geo/GeoShapeUtils.java
Comment (nknize) [reply server/src/main/java/org/opensearch/common/geo/GeoShapeUtils.java]: Rather than put the fix in the utility API I added it directly to the VectorGeoShapeQueryProcessor. This way quantization only happens at query time. I also added a TODO to fix this sandy Lucene experience upstream

**Comment by nknize** [comment]
start gradle check

**Review by mikemccand — APPROVED** [review]

**Comment by nknize** [comment]
start gradle check

**Comment by nknize** [comment]
start gradle check

**Comment by nknize** [comment]
start gradle check

**Comment by rmuir** [comment]
@nknize this one fails with lots of version related test failures. Most interesting from the build log, are the ones that cause class initialization failures, e.g.:

```
org.opensearch.index.reindex.DeleteByQueryConcurrentTests > testConcurrentDeleteByQueriesOnDifferentDocs FAILED
    java.lang.ExceptionInInitializerError
        at org.opensearch.cluster.node.DiscoveryNodeRole.<clinit>(DiscoveryNodeRole.java:227)
        ...

        Caused by:
        java.lang.NullPointerException: Cannot read field "id" because "version" is null
            at org.opensearch.Version.<clinit>(Version.java:114)
            ... 49 more

    java.lang.NoClassDefFoundError: Could not initialize class org.opensearch.Version
```

I'm also seeing wacky racy stuff locally with NoClassDefFoundError for integ tests (in the main branch), and also a unit test failure (~50% of the time) related to versioning (which i have muted locally already).

I think part of the problem is that there is a circular clinit dependency between `Version` and `LegacyESVersion`. E.g. `LegacyESVersion extends Version` but `Version.static {}` refers to `LegacyESVersion`. I think this is causing classloading issues.

Should we revisit/solidify this version stuff before trying to upgrade lucene?

**Comment by nknize** [comment]
> I think part of the problem is that there is a circular clinit dependency between `Version` and `LegacyESVersion`.

That was exactly the problem. I pushed a change so the `LegacyESVersion` fields are loaded before the static initialization block but this "fix" is still just another hack around the hack. It should fix the race conditions, though, so we could move forward and unblock a lucene upgrade (I'd like to get main on Lucene 9.0.0 ASAP to begin using all the cool Vector field stuff).  

It doesn't fix the inherited tech debt though. IMHO the versioning logic needs a complete rewrite (interested? :) ) . The gradle implementation, for example, reads the Version.java as a text file and uses a regex to parse declared versions from the variable names! It's soo hacky.

**Comment by nknize** [comment]
start gradle check

**Comment by rmuir** [comment]
IMO significantly changing this kind of logic almost always results in terrible bugs. So at the least we should reduce risk by pulling these changes away from this unrelated lucene upgrade PR, and get main branch stable?

Also maybe it is worth reconsidering how important is it to really reboot at version 1.0? I'm just throwing out the idea, because continuing with existing numbers means you could throw out LegacyESVersion completely, correct? It would reduce risk even more.

**Comment by nknize** [comment]
> So at the least we should reduce risk by pulling these changes away from this unrelated lucene upgrade PR,

Yeah, I think splitting this into two PRs is important to track this tech debt. I just started doing that to avoid polluting the lucene changes PR w/ the versioning hacks. (looks like the CI infra is down though so we're getting that looked at) 

> continuing with existing numbers means you could throw out LegacyESVersion completely, correct?

Yes. And it would make things a LOT easier; but I think there are non-technical issues carrying forward w/ the versioning that lead to the desire to cleanly rebase the OpenSearch version to 1.0.0.

**Comment by nknize** [comment]
going to let the gradle check test complete before opening the split PR.

**Comment by peterzhuamazon** [comment]
start gradle precommit

**Comment by rmuir** [comment]
I can't tell, but it looks like the last one failed for several reasons:
* version related test failures
* thread blocked on entropy (seeding DRBG random)

Do CI machines have appropriate entropy (either newish kernel with `random.trust_cpu=1` parameter or running `rngd` to seed entropy pools from `RDRAND` insn ?). I'm not sure all these tests and so on are careful about consuming precious entropy.

**Comment by nknize** [comment]
> Do CI machines have appropriate entropy (either newish kernel with `random.trust_cpu=1` parameter or running `rngd` to seed entropy pools from `RDRAND` insn ?)

/cc @bbarani @peterzhuamazon  (perhaps we should have an @infra team to ping the CI maintainers?)

**Comment by rmuir** [comment]
> > continuing with existing numbers means you could throw out LegacyESVersion completely, correct?
> 
> Yes. And it would make things a LOT easier; but I think there are non-technical issues carrying forward w/ the versioning that lead to the desire to cleanly rebase the OpenSearch version to 1.0.0.

But current approach tries to take on too much: reversioning to 1.0, backwards compat with old 7.x versions (without tests). This is too complex and tests/build/code is failing everywhere, we should revert this reversioning!

Non-technical people can be informed that it didn't work out, at least this time. we can always try again at some later time. Gotta keep these PM-types on a short leash, or they will re-label and re-version you every day. They have no idea of e.g. the costs of backwards compatibility support.

**Comment by nknize** [comment]
> backwards compat with old 7.x versions (without tests)

This!  I don't think the rebase to 1.0.0 is so much a concern as is the desire for backwards compatibility. It would be nice to have a clean break altogether and if the community wanted backward, sideways, cross, insideout compatibility it could be looked at in a separate plugin.

**Comment by nknize** [comment]
start gradle check

**Comment by nknize** [comment]
start gradle check

**Comment by nknize** [comment]
start gradle check

**Comment by nknize** [comment]
start gradle check

**Comment by nknize** [comment]
start gradle check

**Comment by nknize** [comment]
start gradle check

**Comment by nknize** [comment]
start gradle check

**Comment by nknize** [comment]
Unable to reproduce these flaky failing tests. I can't tell if these are CI induced or legitimate failures! 

Example from last run:

```
REPRODUCE WITH: ./gradlew ':client:rest-high-level:asyncIntegTest' --tests "org.opensearch.client.ReindexIT.testDeleteByQueryTask" -Dtests.seed=4F5E74A42FAF0A0 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1" -Dtests.locale=ar-BH -Dtests.timezone=America/Pangnirtung -Druntime.java=15
```
```
Dtests.timezone=America/Pangnirtung -Druntime.java=15
  2> java.lang.AssertionError
        at __randomizedtesting.SeedInfo.seed([4F5E74A42FAF0A0:C9242E500657469F]:0)
        at org.junit.Assert.fail(Assert.java:86)
        at org.junit.Assert.assertTrue(Assert.java:41)
        at org.junit.Assert.assertTrue(Assert.java:52)
        at org.opensearch.client.OpenSearchRestHighLevelClientTestCase.lambda$checkTaskCompletionStatus$2(OpenSearchRestHighLevelClientTestCase.java:360)
        at org.opensearch.test.OpenSearchTestCase.assertBusy(OpenSearchTestCase.java:1019)
        at org.opensearch.test.OpenSearchTestCase.assertBusy(OpenSearchTestCase.java:992)
        at org.opensearch.client.ReindexIT.testDeleteByQueryTask(ReindexIT.java:327)

**Pull request merged**
